Text/XHtml/Internals.hs

```diff
@@ -272,79 +317,83 @@ prettyHtmlFragment =
-
-renderHtml' :: Int -> HtmlElement -> ShowS
-renderHtml' _ (HtmlString str) = (++) str
+    = if isValidHtmlITag name && isNoHtml html
+      then renderTag True name (attrs []) ""
+      else mconcat
+        [ renderTag False name (attrs []) ""
+        , foldMap showHtml' (getHtmlElements html)
+        , renderEndTag name ""
+        ]
+
+renderHtml' :: Int -> HtmlElement -> Builder
+renderHtml' _ (HtmlString str) = str
 renderHtml' n (HtmlTag
               { markupTag = name,
                 markupContent = html,
                 markupAttrs = attrs })
-      = if isNoHtml html && elem name validHtmlITags
-        then renderTag True name attrs (nl n)
-        else (renderTag False name attrs (nl n)
-             . foldr (.) id (map (renderHtml' (n+2)) (getHtmlElements html))
-             . renderEndTag name (nl n))
+      = if isValidHtmlITag name && isNoHtml html
+        then renderTag True name (attrs []) (nl n)
+        else renderTag False name (attrs []) (nl n)
+          <> foldMap (renderHtml' (n+2)) (getHtmlElements html)
+          <> renderEndTag name (nl n)
     where
-      nl n' = "\n" ++ replicate (n' `div` 8) '\t'
-              ++ replicate (n' `mod` 8) ' '
+      nl n' = "\n" <> Sem.stimes (n' `div` 8) (charUtf8 '\t')
+              <> Sem.stimes (n' `mod` 8) (charUtf8 ' ')
 
 
 prettyHtml' :: HtmlElement -> [String]
-prettyHtml' (HtmlString str) = [str]
+prettyHtml' (HtmlString str) = [builderToString str]
 prettyHtml' (HtmlTag
               { markupTag = name,
                 markupContent = html,
                 markupAttrs = attrs })
-      = if isNoHtml html && elem name validHtmlITags
+      = if isValidHtmlITag name && isNoHtml html
         then
-         [rmNL (renderTag True name attrs "" "")]
+         [rmNL (renderTag True name (attrs []) "")]
         else
-         [rmNL (renderTag False name attrs "" "")] ++
+         [rmNL (renderTag False name (attrs []) "")] ++
           shift (concat (map prettyHtml' (getHtmlElements html))) ++
-         [rmNL (renderEndTag name "" "")]
+         [rmNL (renderEndTag name "")]
   where
       shift = map (\x -> "   " ++ x)
-      rmNL = filter (/= '\n')
+      rmNL = filter (/= '\n') . builderToString
 
 
 -- | Show a start tag
 renderTag :: Bool       -- ^ 'True' if the empty tag shorthand should be used
-          -> String     -- ^ Tag name
+          -> Builder     -- ^ Tag name
           -> [HtmlAttr] -- ^ Attributes
-          -> String     -- ^ Whitespace to add after attributes
-          -> ShowS
-renderTag empty name attrs nl r
-      = "<" ++ name ++ shownAttrs ++ nl ++ close ++ r
+          -> Builder     -- ^ Whitespace to add after attributes
+          -> Builder
+renderTag empty name attrs nl
+      = "<" <> name <> shownAttrs <> nl <> close
   where
       close = if empty then " />" else ">"
 
-      shownAttrs = concat [' ':showPair attr | attr <- attrs ]
+      shownAttrs = foldMap (\attr -> charUtf8 ' ' <> showPair attr) attrs
```

Comment:
In my equivalence testing, the attributes are showing up in reverse order. Weird. The difference should be pretty small.

`concat [ k x | x <- xs ]` is equal to `concat $ map k xs`, which is `concatMap k xs`, and `concatMap = foldMap` when `m ~ [b]`, as it is here, so the real diff is 

```diff
-foldMap (\attr -> ' ' : showPair attr) attrs
+foldMap (\attr -> charUtf8 ' ' <> showPair attr) attrs
```

So the rendering isn't the issue.